[docs] Update Unit testing guide

amandeepmittal opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Fixes ENG-12679

The `react-test-renderer` library is deprecated by React team and we need to update this guide to provide a solution for writing unit and snapshot tests using Jest.

# How

<!--
How did you build this feature or fix this bug and why?
-->

Changes:
- Rename the guide to "Unit testing with Jest" to reflect the right information. Jest is more than a Unit testing framework. It also allows snapshot testing.
- Fix grammatical error in the description
- Update the guide's installation instructions based on the default template structure, `jest-expo` updates. Also, manual installation and configuration instructions should be provided when the Expo project is not created with the default template.
- Add info to install React Native Testing Library and update the following instructions and examples to demonstrate that library's usage for writing unit and snapshot tests
- Update structure for organizing tests section to write as per the default template
- Update Code Coverage snippet to exclude other expo generated files and directories
- Add a link to React Native Testing Library documentation (external)
- Add a link to e2e tests guide (internal)
- Use TypeScript as the primary language for all code examples

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

Run docs locally and visit: http://localhost:3002/develop/testing or see **Preview**.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

amandeepmittal committed:
[docs] Update Testing guide (previously called Unit testing)

keith-kurak commented on docs/pages/develop/testing.mdx:
```diff
@@ -0,0 +1,319 @@
+---
+title: Testing with Jest
+sidebar_title: Testing
+description: Learn how to set up and configure the jest-expo library to write unit and snapshot tests for a project.
```
```suggestion
description: Learn how to set up and configure the jest-expo library to write unit and snapshot tests for a project with Jest.
```
Maybe too redundant, was just thinking about how it's specifically about Jest, and how there's other ways to write unit tests

keith-kurak commented on docs/pages/develop/testing.mdx:
```diff
@@ -0,0 +1,319 @@
+
+[Jest](https://jestjs.io) is the most widely used unit and snapshot JavaScript testing framework. In this guide, you will learn how to set up Jest in your project, write a unit test, write a snapshot test, and best practices for structuring your tests when using Jest with React Native.
+
+You will also use the [`jest-expo`](https://github.com/expo/expo/tree/main/packages/jest-expo) library, which is a Jest preset and mocks the native part of the Expo SDK and handles most of the configuration.
```
```suggestion
You will also use the [`jest-expo`](https://github.com/expo/expo/tree/main/packages/jest-expo) library, which is a Jest preset that mocks the native part of the Expo SDK and handles most of the configuration needed for your Expo project.
```

keith-kurak commented on docs/pages/develop/testing.mdx:
```diff
@@ -0,0 +1,319 @@
+
+</Tabs>
+
+Jest has various configuration options. The above configuration covers the majority of your needs. However, you can always add to this pattern list. For more details, see [Configuring Jest](https://jestjs.io/docs/configuration).
```
```suggestion
Jest has many configuration option, but the above configuration should cover most of your needs. However, you can always add to this pattern list. For more details, see [Configuring Jest](https://jestjs.io/docs/configuration).
```

amandeepmittal replied:
I think splitting the sentence is okay here. we don't have to add pause and introduce contradiction. Also, "options" is the plural form that references those multiple configs that Jest provides.

amandeepmittal replied:
I've taken the simplification parts from your suggestions. Let me know if you still think it's not good.

keith-kurak replied:
oops, I meant to make that plural!

keith-kurak reviewed (changes requested):
Looks good! A few proposed tweaks and perhaps one big question:

Since this is our only / main "unit testing" page, do we lose something in terms of searchability by renaming it from "unit testing?" Even though "testing with Jest" is more accurate, we're not really suggesting any other options for unit testing. 

I'm not sure of the answer, just curious. Maybe I'm alone in this, but I don't find it all that bad to categorize snapshot tests under unit tests. Maybe they're not technically the same, but they both usually test irreducible "units".

amandeepmittal commented on docs/pages/develop/testing.mdx:
```diff
@@ -0,0 +1,319 @@
+  "preset": "jest-expo",
+  /* @info */
+  "transformIgnorePatterns": [
+    "node_modules/(?!((jest-)?react-native|@react-native(-community)?)|expo(nent)?|@expo(nent)?/.*|@expo-google-fonts/.*|react-navigation|@react-navigation/.*|@unimodules/.*|unimodules|@sentry/react-native|native-base|react-native-svg)"
```
I can remove this from here. However, I verified this info from the `jest-preset` library itself: https://github.com/expo/expo/blob/10d6426b2f0a3ecd54ff888e6d2758de35ffe743/packages/jest-expo/jest-preset.js#L57.

amandeepmittal replied:
Unimodules part is now removed from the doc.

keith-kurak replied:
it's harmless, so if you think it's better for consistency, we can leave it there. Though I wonder if we can safely remove it from jest-presets. Doctor will yell at you if you import unimodules :-).

amandeepmittal replied:
Thanks for the addtional context about that. I've removed it from the guide :)

amandeepmittal committed:
Apply feedback from Keith

amandeepmittal commented:
> Since this is our only / main "unit testing" page, do we lose something in terms of searchability by renaming it from "unit testing?" Even though "testing with Jest" is more accurate, we're not really suggesting any other options for unit testing.

When re-writing the title of this page, I thought it could act as a kind of landing page too (since there's a section now called "Additional information," which links to other testing guides).

I thought it was technically appropriate to convey that this guide provides info about setting up testing with Jest and  snapshot testing, that's why I changed the title.

Not sure about the last part of your sentence. Should we include more options for unit testing? If yes, then maybe we can add links here and create a separate section about that under "Guides" so that this doc doesn't have to act as a reference page (most docs under Home are oriented towards a beginner/get-started persona of a user so that it doesn't get overwhelming).

re: About searchability
Yeah, the title change will affect the Algolia search. I think that for SEO search, the description of the page uses keywords like "unit" and "snapshot," so that should cover it.

If you think we should not lose "Unit testing" from searchability internally, I can also modify the title to "Unit testing with Jest" and keep the URL of the page unmodified.

amandeepmittal committed:
Update testing.mdx

keith-kurak commented:
@amandeepmittal I like "Unit Testing with Jest" from a searchability perspective. I don't believe we should include other testing libraries, was just thinking of the double-duty this is performing as a Jest guide, but as the resource for unit and other-unit-ish testing.

amandeepmittal committed:
Apply feedback from Keith

amandeepmittal commented:
@keith-kurak, I've updated the page title to "Unit testing with Jest". Also, I reverted the URL to `/unit-testing/` this guide (I don't think we need to change that anymore based on the title of this guide). Removed the redirects accordingly that were initially introduced in this PR.
 
![CleanShot 2024-08-07 at 00 39 20@2x](https://github.com/user-attachments/assets/326c9e24-7f51-4b55-b287-beee9b87a6a5)

keith-kurak reviewed (approved):
LGTM!
